src/store_client.cc

```diff
@@ -499,58 +515,126 @@ store_client::fileRead()
+        return fail();
     }
 
-    if (len > 0 && rep && entry->mem_obj->inmem_lo == 0 && entry->objectLen() <= (int64_t)Config.Store.maxInMemObjSize && Config.onoff.memory_cache_disk) {
-        storeGetMemSpace(len);
+    assert(lastRead_.data == buf);
+    lastRead_.length = lastIoResult;
+
+    parsingBuffer->appended(buf, lastIoResult);
+
+    maybeWriteFromDiskToMemory(lastRead_);
+    handleBodyFromDisk();
+}
+
+/// de-serializes HTTP response (partially) read from disk storage
+void
+store_client::handleBodyFromDisk()
+{
+    // We cannot de-serialize on-disk HTTP response without MemObject because
+    // without MemObject::swap_hdr_sz we cannot know where that response starts.
+    Assure(entry->mem_obj);
+    Assure(entry->mem_obj->swap_hdr_sz > 0);
+
+    if (!answeredOnce()) {
+        // All on-disk responses have HTTP headers. First disk body read(s)
+        // include HTTP headers that we must parse (if needed) and skip.
+        const auto haveHttpHeaders = entry->mem_obj->baseReply().pstate == Http::Message::psParsed;
+        if (!haveHttpHeaders && !parseHttpHeadersFromDisk())
+            return;
+        skipHttpHeadersFromDisk();
+    }
+
+    // no callback(): we handled negative/zero lastIoResult ourselves
+    noteNews();
+}
+
+/// Adds HTTP response data loaded from disk to the memory cache (if
+/// needed/possible). The given part may contain portions of HTTP response
+/// headers and/or HTTP response body.
+void
+store_client::maybeWriteFromDiskToMemory(const StoreIOBuffer &httpResponsePart)
+{
+    // XXX: Reject [memory-]uncachable/unshareable responses instead of assuming
+    // that any HTTP response loaded from disk should be written to MemObject's
+    // data_hdr (and that it may purge already cached entries). Cachability
+    // decision(s) should be made outside (and obeyed by) this low-level code.
+    if (httpResponsePart.length && entry->mem_obj->inmem_lo == 0 && entry->objectLen() <= (int64_t)Config.Store.maxInMemObjSize && Config.onoff.memory_cache_disk) {
+        storeGetMemSpace(httpResponsePart.length);
+        // XXX: The "recheck" below is not needed because nobody can purge
+        // mem_hdr bytes of a locked entry, and we do lock our entry. Moreover,
+        // inmem_lo offset itself should not be relevant to appending new bytes.
         // recheck for the above call may purge entry's data from the memory cache
         if (entry->mem_obj->inmem_lo == 0) {
-            /* Copy read data back into memory.
-             * copyInto.offset includes headers, which is what mem cache needs
-             */
-            if (copyInto.offset == entry->mem_obj->endOffset()) {
-                entry->mem_obj->write(StoreIOBuffer(len, copyInto.offset, copyInto.data));
-            }
+            // XXX: This code assumes a non-shared memory cache.
+            if (httpResponsePart.offset == entry->mem_obj->endOffset())
```

Comment:
In one of my tests this condition returned false because was httpResponsePart.offset = entry->mem_obj->endOffset() + swap_hdr_sz. Should httpResponsePart.offset include swap_hdr_sz?

Reply:
> Should httpResponsePart.offset include swap_hdr_sz?

No, httpResponsePart should not because HTTP response does not include swap metadata. One could argue that the name of that variable is wrong, but, AFAIK, MemObject::endOffset() is also supposed to be just about the HTTP response -- MemObject does not store any swap metadata.

However, I think there is a bug here: The caller is given this function something other than httpResponsePart because lastRead is a disk I/O buffer (i.e. its offset does include swap_hdr_sz). I will fix this shortly.

Reply:
Before this fix, I noticed a strange behavior in my test: if an entry was cached on rock disk an requested, only the initial 16K part was copied into the local memory - subsequent requests still got the rest of the entry from the disk. Anyway, the fix seems to resolve this problem.

Reply:
Glad to hear the problem appears to be gone. Unfortunately, that particular commit had some unwanted changes. Please use 78e2f8d or later.